python-frontend/src/test/java/org/sonar/python/types/TypeInferenceTest.java

```diff
@@ -276,6 +276,23 @@ public void conditional_expressions() {
       "c").type()).isEqualTo(anyType());
   }
 
+  @Test
+  public void logical_expressions() {
+    assertThat(lastExpression("not 42").type()).isEqualTo(BOOL);
+    assertThat(lastExpression("42 or 43").type()).isEqualTo(INT);
+    assertThat(lastExpression("42 and 43").type()).isEqualTo(INT);
+    assertThat(lastExpression("42 or ''").type()).isEqualTo(or(INT, STR));
+    assertThat(lastExpression("42 or xxx").type()).isEqualTo(anyType());
+    assertThat(lastExpression("42 or True or ''").type()).isEqualTo(or(or(INT, STR), BOOL));
+
+    assertThat(lastExpressionInFunction(
+      "for i in range(3):",
+      "  if   i > 1: c = b",
+      "  elif i > 0: b = 42 or a",
+      "  else:       a = ''",
+      "c").type()).isEqualTo(or(INT, STR));
```

Comment:
Here we test both:
* something which is specific to logical expressions: `BinaryExpressionImpl#typeDependencies()`
* the algorithm to propagate types between expressions.

I think that we should have only 1 test for the second point and test the various implementations of `typeDependencies()` explicitly.